### PR TITLE
TYP: annotate _data

### PR DESCRIPTION
**Pull request by jbrockmendel**
NB: this does induce a small behavior change in that `Index` will not have a `_data` attribute until `__init__/__new__`.

I found in #29561 that getting `_data` annotated is necessary before we can annotate most of `core.indexes`.

I'm also not sure what to do about MultiIndex, suggestions welcome.

### CODE REVIEW

**Comment by jreback** [comment]
lgtm. @simonjayhawkins @WillAyd

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ class DatetimeIndexOpsMixin(ExtensionOpsMixin):
     common ops mixin to support a unified interface datetimelike Index
     """
 
-    _data = None
```
Comment (WillAyd) [inline pandas/core/indexes/datetimelike.py]: Were these solely set to `None` before for annotation purposes? I think need to be careful in changing this if the intent was solely solely for that purpose

Comment (jbrockmendel) [reply pandas/core/indexes/datetimelike.py]: these were here pre-annotations, so I think it was so that the attribute always exists

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -230,7 +230,7 @@ def _outer_indexer(self, left, right):
         return libjoin.outer_join_indexer(left, right)
 
     _typ = "index"
-    _data = None
+    _data: Union[ExtensionArray, np.ndarray]
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py]: you can now remove the assert and comment added in #29475 on L657?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: yes, updated

Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: updated+green

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**